api/v1alpha4/virtualmachinegroup_publishrequest_types.go

```diff
@@ -0,0 +1,196 @@
+	// +optional
+
+	// Source is the name of the VirtualMachineGroup to be published.
+	//
+	// If this value is omitted then the publication controller checks to
+	// see if there is a resource with the same name as this
+	// VirtualMachineGroupPublishRequest resource. If such a resource exists,
+	// then it is the source of the publication.
+	Source string `json:"source,omitempty"`
+
+	// +optional
+
+	// VirtualMachines is a list of the VirtualMachine objects from the source
+	// VirtualMachineGroup that are included in this publish request.
+	//
+	// If omitted, this field defaults to the names of all of the VMs currently
+	// a member of the group, either directly or indirectly via a nested group.
+	VirtualMachines []string `json:"virtualMachines,omitempty"`
+
+	// +optional
+
+	// Target is the name of the ContentLibrary resource to which the
+	// VirtualMachines from the VirtualMachineGroup should be published.
+	//
+	// When this value is omitted, the controller attempts to identify the
+	// target location by matching a ContentLibrary resource with the label
+	// "imageregistry.vmware.com/default".
+	//
+	// Please note that while optional, if a VirtualMachineGroupPublishRequest
+	// sans target information is applied to a namespace without a default
+	// publication target, then the VirtualMachineGroupPublishRequest resource
+	// will be marked in error.
+	Target string `json:"target,omitempty"`
+
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+
+	// TTLSecondsAfterFinished is the time-to-live duration for how long this
+	// resource will be allowed to exist once the publication operation
+	// completes. After the TTL expires, the resource will be automatically
+	// deleted without the user having to take any direct action.
+	// This will be passed into each VirtualMachinePublishRequestSpec.
+	//
+	// If this field is unset then the request resource will not be
+	// automatically deleted. If this field is set to zero then the request
+	// resource is eligible for deletion immediately after it finishes.
+	TTLSecondsAfterFinished *int64 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+
+// VirtualMachineGroupPublishRequestStatus defines the observed state of a
+// VirtualMachineGroupPublishRequest.
+type VirtualMachineGroupPublishRequestStatus struct {
+	// +optional
+
+	// Source is the name of the published VirtualMachineGroup.
+	Source string `json:"source,omitempty"`
+
+	// +optional
+
+	// Target is the name of the ContentLibrary to which the group is published.
+	Target string `json:"target,omitempty"`
+
+	// +optional
+
+	// CompletionTime represents when the request was completed. It is not
+	// guaranteed to be set in happens-before order across separate operations.
+	// It is represented in RFC3339 form and is in UTC.
+	//
+	// The value of this field should be equal to the value of the
+	// LastTransitionTime for the status condition Type=Complete.
+	CompletionTime metav1.Time `json:"completionTime,omitempty"`
```

Comment:
Should we use pointer types for `CompletionTime` and `StartTime` fields to make the `omitempty` JSON tags effective, as it doesn't work properly with struct values?

Reply:
No, because it's not really a struct. Check out the encoder/decoder logic.